### PR TITLE
parser+typechecker: Accept fat arrow with explicit return type

### LINKED ISSUE #331: Confusing error when defining => functions with a -> type
Functions defined using `=>` have their return type inferred automatically, and are not allowed to have it stated explicitly with `-> Foo`. However, the error you get is not at all clear. With this example code:

```jakt
class Limited {
    public function get_secret() -> String => "Shhhh! Don't tell anyone!"
}

function main() {
    Limited::get_secret();
}
```

...we get this error:

```
Error: expected field
----
 6 | } 
----
```

A better error would tell you that the method definition is invalid because of the `-> String`, and point to it.

EDIT: Alternatively, maybe explicit return types should be allowed and override the inferred one? Either way, this is a confusing error.

### PR DESCRIPTION
**Pull request by cybergsus**
Now Jakt accepts specifying a return type explicitly for functions using a fat arrow to declare their block as a single expression with
a return, and makes sure that the type of the expression matches with the explicit type:

```jakt
class S {
  function get_secret() -> i32 => "World"
}
```

would result in a `Type mismatch` error with `i32` and `String`, but

```jakt
function secret() -> i32 => 42
```

won't throw any errors.

Works for both methods and standalone functions.
Will close #331.

### CODE REVIEW

File: tests/typechecker/function-return-type-with-fat-arrow.jakt [hunk 1]
```diff
@@ -0,0 +1,3 @@
+class S {
+  public function get_value() -> i32 => "Hello"
```
Comment (linusg) [inline tests/typechecker/function-return-type-with-fat-arrow.jakt:2]: 4 spaces please

File: src/parser.rs [hunk 2]
```diff
@@ -1495,48 +1497,43 @@ pub fn parse_function(
...
+                    Some(Span {
+                        file_id,
+                        start,
+                        end: tokens[*index - 1].span.end,
+                    })
+                } else {
+                    None
+                };
+
+                // accept an (optional)  fat arrow
```
Comment (linusg) [inline src/parser.rs:1524]: double space in comment here

**Review by linusg — CHANGES_REQUESTED** [review]

File: src/parser.rs (hunk 2 above)
Comment (cybergsus) [reply src/parser.rs]: That code goes through automatic formatter. I can't enforce that manually.

Comment (cybergsus) [reply src/parser.rs]: Nevermind, I could enforce that manually.

**Review by alimpfard — COMMENTED** [review]
Just a (repetitive) nitpick

File: src/parser.rs [hunk 3]
```diff
@@ -1495,48 +1497,43 @@ pub fn parse_function(
...
+                    Some(Span {
+                        file_id,
+                        start,
+                        end: tokens[*index - 1].span.end,
+                    })
+                } else {
+                    None
+                };
+
+                //  Accept an (optional)  fat arrow
```
[the remark this answers is not in the document]
Comment (cybergsus) [reply src/parser.rs:1524]: oh I thought the space was needed

[the remark this answers is not in the document]
Comment (AtkinsSJ) [reply src/parser.rs]: Linus was talking about the double space between "(optional)" and "fat", which shouldn't be there. :^)

**Comment by awesomekling** [comment]
Seems like we could also have a test where the parse succeeds instead of fails :)

**Comment by cybergsus** [comment]
I'll create a `.out` file as well with a sample output from a sample program that uses this syntax then

**Comment by cybergsus** [comment]
Thanks! <3

**Pull request merged**